ttl/ttlworker/job.go

```diff
@@ -0,0 +1,180 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttlworker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/ttl/cache"
+	"github.com/pingcap/tidb/ttl/session"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const updateJobCurrentStatusTemplate = "UPDATE mysql.tidb_ttl_table_status SET current_job_status = '%s' WHERE table_id = %d AND current_job_status = '%s' AND current_job_id = '%s'"
+const finishJobTemplate = "UPDATE mysql.tidb_ttl_table_status SET last_job_id = current_job_id, last_job_start_time = current_job_start_time, last_job_finish_time = '%s', last_job_ttl_expire = current_job_ttl_expire, last_job_summary = '%s', current_job_id = NULL, current_job_owner_id = NULL, current_job_owner_hb_time = NULL, current_job_start_time = NULL, current_job_ttl_expire = NULL, current_job_state = NULL, current_job_status = NULL, current_job_status_update_time = NULL WHERE table_id = %d AND current_job_id = '%s'"
+const updateJobStateTemplate = "UPDATE mysql.tidb_ttl_table_status SET current_job_state = '%s' WHERE table_id = %d AND current_job_id = '%s'"
+
+func updateJobCurrentStatusSQL(tableID int64, oldStatus cache.JobStatus, newStatus cache.JobStatus, jobID string) string {
+	return fmt.Sprintf(updateJobCurrentStatusTemplate, newStatus, tableID, oldStatus, jobID)
+}
+
+func finishJobSQL(tableID int64, finishTime time.Time, summary string, jobID string) string {
+	return fmt.Sprintf(finishJobTemplate, finishTime.Format(timeFormat), summary, tableID, jobID)
+}
+
+func updateJobState(tableID int64, currentJobID string, currentJobState string) string {
+	return fmt.Sprintf(updateJobStateTemplate, currentJobState, tableID, currentJobID)
+}
+
+type ttlJob struct {
+	id string
+
+	ctx    context.Context
+	cancel func()
+
+	createTime time.Time
+
+	tbl *cache.PhysicalTable
+
+	tasks                   []*ttlScanTask
+	taskIter                int
+	finishedScanTaskCounter int
+	scanTaskErr             error
+
+	// status is the only field which should be protected by a mutex, as `Cancel` may be called at any time, and will
+	// change the status
+	statusMutex sync.Mutex
+	status      cache.JobStatus
+
+	statistics *ttlStatistics
+}
+
+// changeStatus updates the state of this job
+func (job *ttlJob) changeStatus(ctx context.Context, se session.Session, status cache.JobStatus) error {
+	job.statusMutex.Lock()
+	defer job.statusMutex.Unlock()
+
+	job.status = status
+
+	_, err := se.ExecuteSQL(ctx, updateJobCurrentStatusSQL(job.tbl.ID, job.status, status, job.id))
```

Comment:
> I think it's better to ExecuteSQL outside the lock.

Good. Fixed.

> Actually, I think we should make sure everything goes as expected even if the SQL is failed to execute which leads to a inconsistency between memory and persisted data.

Now, I hand over the choice from this function to the caller. For most situation (e.g. timeout, change to running), the error will only be logged. The scheduler doesn't depend on these status.

The only exception about `Cancel`. The cancel status is used to assert whether the job has been cancelled, so I return the error (and it will finally be returned to the sql response) to the user. But as we still modified the memory, the job will be cancelled as well. Maybe I could ignore this error either?